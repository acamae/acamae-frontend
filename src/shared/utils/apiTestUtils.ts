import apiService from '@shared/services/axiosService';

/**
 * Represents the structured body of an API response, as defined by the application.
 */
export interface ApiResponseBody<TData = null> {
  success: boolean;
  data?: TData;
  message?: string;
  errors?: { field: string; errorKey: string }[];
}

// Simula la estructura de error de Axios para los mocks
export interface MockAxiosError extends Error {
  isAxiosError: boolean;
  response: {
    data: unknown;
    status: number;
  };
}

// Esta interfaz se exporta para que los tests puedan tipar sus par√°metros si lo desean
export interface SimulateApiCallParams<TResponseData = unknown, TRequestData = unknown> {
  method: 'get' | 'post' | 'put' | 'delete';
  url: string;
  requestData?: TRequestData;
  httpStatusCode: number;
  apiResponseBody: ApiResponseBody<TResponseData>;
  networkError?: Error; // Para simular errores donde no hay respuesta HTTP
}

export const setupMockApiServiceCall = <TResponseData = unknown, TRequestData = unknown>(
  params: SimulateApiCallParams<TResponseData, TRequestData>
) => {
  const {
    method,
    url: expectedUrl,
    requestData: expectedRequestData,
    httpStatusCode,
    apiResponseBody,
    networkError,
  } = params;

  const serviceMethod = (apiService as jest.Mocked<typeof apiService>)[method];

  if (networkError) {
    serviceMethod.mockImplementation(async (url: string, requestPayload?: unknown) => {
      if (url !== expectedUrl) {
        return Promise.reject(
          new Error(
            `Mocked ${method} (network error sim) called with unexpected URL: ${url}. Expected: ${expectedUrl}`
          )
        );
      }
      if (
        (method === 'post' || method === 'put') &&
        expectedRequestData &&
        JSON.stringify(requestPayload) !== JSON.stringify(expectedRequestData)
      ) {
        return Promise.reject(
          new Error(
            `Mocked ${method} (network error sim) to ${url} called with unexpected data. Expected: ${JSON.stringify(expectedRequestData)}, Got: ${JSON.stringify(requestPayload)}`
          )
        );
      }
      return Promise.reject(networkError);
    });
    return;
  }

  serviceMethod.mockImplementation(async (url: string, requestPayload?: unknown) => {
    if (url !== expectedUrl) {
      return Promise.reject(
        new Error(
          `Mocked ${method} (status ${httpStatusCode}) called with unexpected URL: ${url}. Expected: ${expectedUrl}`
        )
      );
    }
    if (
      (method === 'post' || method === 'put') &&
      expectedRequestData &&
      JSON.stringify(requestPayload) !== JSON.stringify(expectedRequestData)
    ) {
      return Promise.reject(
        new Error(
          `Mocked ${method} (status ${httpStatusCode}) to ${url} called with unexpected data. Expected: ${JSON.stringify(expectedRequestData)}, Got: ${JSON.stringify(requestPayload)}`
        )
      );
    }

    if (httpStatusCode >= 200 && httpStatusCode < 300) {
      return Promise.resolve({
        data: apiResponseBody,
        status: httpStatusCode,
      });
    } else if (httpStatusCode >= 300 && httpStatusCode < 600) {
      const error: MockAxiosError = Object.assign(
        new Error(`Request failed with status code ${httpStatusCode}`),
        {
          isAxiosError: true,
          response: {
            data: apiResponseBody,
            status: httpStatusCode,
          },
        }
      );
      return Promise.reject(error);
    } else {
      return Promise.reject(
        new Error(
          `Unsupported HTTP status code ${httpStatusCode} in mock setup for URL ${expectedUrl}.`
        )
      );
    }
  });
};

export const promiseMock = (shouldResolve = true) => {
  return jest.fn(() =>
    shouldResolve ? Promise.resolve() : Promise.reject(new Error('mock error'))
  );
};
