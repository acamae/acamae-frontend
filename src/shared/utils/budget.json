{
  "categories": {
    "performance": 90,
    "accessibility": 90,
    "best-practices": 90,
    "seo": 90
  },
  "metrics": {
    "largest-contentful-paint": [2500, 4000],
    "first-input-delay": [100, 300],
    "cumulative-layout-shift": [0.1, 0.25],
    "first-contentful-paint": [1800, 3000],
    "interactive": [3800, 7300],
    "total-blocking-time": [200, 600],
    "speed-index": [3400, 5800]
  },
  "timing": {
    "server-response-time": 600
  }
}
