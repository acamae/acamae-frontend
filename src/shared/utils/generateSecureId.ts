/**
 * Genera un ID criptográficamente seguro
 * Usa crypto.randomUUID() si está disponible, o crypto.getRandomValues() como alternativa
 *
 * @param dependencies Dependencias externas inyectables para testing
 */
export function generateSecureId(
  dependencies = {
    cryptoObj: typeof crypto !== 'undefined' ? crypto : undefined,
    dateNow: Date.now,
    processHrtime: typeof process !== 'undefined' ? process.hrtime : undefined,
  }
): string {
  const { cryptoObj, dateNow, processHrtime } = dependencies;

  try {
    // Intentar usar crypto.randomUUID() (disponible en Node.js 14.17+ y navegadores modernos)
    if (cryptoObj && typeof cryptoObj.randomUUID === 'function') {
      return cryptoObj.randomUUID().replace(/-/g, '').slice(0, 9);
    }

    // Alternativa con crypto.getRandomValues() (disponible en Node.js y navegadores)
    if (cryptoObj && typeof cryptoObj.getRandomValues === 'function') {
      const array = new Uint8Array(9);
      cryptoObj.getRandomValues(array);
      return Array.from(array, byte => byte.toString(36))
        .join('')
        .slice(0, 9);
    }

    // Última opción: usar timestamp con process.hrtime para entropía adicional
    // Esto no es criptográficamente seguro pero proporciona mejor entropía que Math.random()
    const timestamp = dateNow();
    const hrtime = processHrtime ? processHrtime.bigint().toString(36) : timestamp.toString(36);
    return `${timestamp}_${hrtime}`.slice(-9);
  } catch {
    // Si todos los métodos fallan, usar un fallback basado en timestamp
    const fallbackTimestamp = dateNow();
    return `${fallbackTimestamp}`.slice(-9);
  }
}
