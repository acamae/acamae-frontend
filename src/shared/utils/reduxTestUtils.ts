import { configureStore } from '@reduxjs/toolkit';

import authReducer, { initialAuthState } from '@application/state/slices/authSlice';
import sessionTimerReducer, {
  initialSessionTimerState,
} from '@application/state/slices/sessionTimerSlice';
import type { RootState } from '@domain/types/redux';

export function makeTestStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: {
      auth: authReducer,
      sessionTimer: sessionTimerReducer,
    },
    preloadedState: {
      auth: initialAuthState,
      sessionTimer: initialSessionTimerState,
      ...preloadedState,
    } as RootState,
    middleware: getDefaultMiddleware => getDefaultMiddleware({ serializableCheck: false }),
  });
}
