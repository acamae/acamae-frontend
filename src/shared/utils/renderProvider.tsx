import { configureStore, Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import rootReducer from '@application/state/rootReducer';
import { RootState } from '@domain/types/redux';
import i18n from '@infrastructure/i18n';

type TestProviderOptions = {
  preloadedState?: Partial<RootState>;
  store?: Store;
  route?: string;
  i18nInstance?: typeof i18n;
  withRouter?: boolean;
  withI18n?: boolean;
};

export function createTestProviderFactory(defaults: Partial<TestProviderOptions> = {}) {
  return function renderWithProviders(ui: React.ReactElement, options: TestProviderOptions = {}) {
    const {
      preloadedState = {},
      store = configureStore({ reducer: rootReducer, preloadedState }),
      route = '/',
      i18nInstance = i18n,
      withRouter = true,
      withI18n = true,
    } = { ...defaults, ...options };

    let tree = <Provider store={store}>{ui}</Provider>;
    if (withI18n) {
      tree = <I18nextProvider i18n={i18nInstance}>{tree}</I18nextProvider>;
    }
    if (withRouter) {
      tree = <MemoryRouter initialEntries={[route]}>{tree}</MemoryRouter>;
    }
    return render(tree);
  };
}
