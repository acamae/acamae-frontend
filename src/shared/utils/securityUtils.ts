import { ThrottleConfig } from '@shared/constants/security';

export interface ThrottleState {
  lastSubmission: number;
  attemptCount: number;
  windowStart: number;
  isBlocked: boolean;
}

/**
 * Safe function to get environment variables with fallback
 * This prevents errors when process.env is not available during module loading
 */
const getEnvVar = (key: string, defaultValue: string): number => {
  try {
    const envValue = process.env[key];
    // Check if the environment variable exists and is not undefined
    if (envValue !== undefined && envValue !== null) {
      return Number(envValue);
    }
    return Number(defaultValue);
  } catch {
    return Number(defaultValue);
  }
};

class SecurityThrottleService {
  private readonly throttleStates: Map<string, ThrottleState> = new Map();
  private readonly defaultConfig: ThrottleConfig = {
    delay: getEnvVar('REACT_APP_THROTTLE_DELAY_DEFAULT', '4000'), // Wait 4 seconds every attempt
    maxAttempts: getEnvVar('REACT_APP_THROTTLE_MAX_ATTEMPTS_DEFAULT', '10'), // Maximum number of attempts
    timeWindow: getEnvVar('REACT_APP_THROTTLE_TIME_WINDOW_DEFAULT', '300000'), // Within time window in milliseconds (5 minutes)
    persistInClient: false,
  };
  private readonly storageKey = 'acamae-throttle-states';

  constructor() {
    this.loadPersistedStates();
  }

  /**
   * Load persisted states from localStorage
   */
  private loadPersistedStates(): void {
    try {
      const stored = localStorage.getItem(this.storageKey);
      if (stored) {
        const data = JSON.parse(stored);
        const now = Date.now();

        // Load only non-expired states
        Object.entries(data).forEach(([actionId, state]) => {
          const throttleState = state as ThrottleState;
          // Check if the state has not expired
          if (now - throttleState.windowStart < 300000) {
            // 5 minutos máximo
            this.throttleStates.set(actionId, throttleState);
          }
        });
      }
    } catch {
      // If there's an error loading, clean corrupted storage
      localStorage.removeItem(this.storageKey);
    }
  }

  /**
   * Persist critical states in localStorage
   */
  private persistCriticalStates(): void {
    try {
      const persistentStates: Record<string, ThrottleState> = {};

      this.throttleStates.forEach((state, actionId) => {
        // Only persist states that require persistence
        if (this.shouldPersistAction(actionId)) {
          persistentStates[actionId] = state;
        }
      });

      if (Object.keys(persistentStates).length > 0) {
        localStorage.setItem(this.storageKey, JSON.stringify(persistentStates));
      } else {
        localStorage.removeItem(this.storageKey);
      }
    } catch (error) {
      // Error persisting - continue without persistence
      console.warn('Failed to persist throttle states:', error);
    }
  }

  /**
   * Check if an action should be persisted based on its configuration
   */
  private shouldPersistAction(actionId: string): boolean {
    // Check if the actionId corresponds to a persistent configuration
    return (
      actionId.includes('login-form') ||
      actionId.includes('register-form') ||
      actionId.includes('forgot-password-form') ||
      actionId.includes('reset-password-form') ||
      actionId.includes('email-verification-resend-form') ||
      actionId.includes('critical-')
    );
  }

  /**
   * Check if a configuration requires persistence
   */
  private requiresPersistence(config: Partial<ThrottleConfig>): boolean {
    return config.persistInClient === true;
  }

  /**
   * Check if an action is allowed based on throttling
   */
  canPerformAction(actionId: string, config?: Partial<ThrottleConfig>): boolean {
    const finalConfig = { ...this.defaultConfig, ...config };
    const now = Date.now();
    const state = this.throttleStates.get(actionId);

    // Si delay es 0, permitir siempre la acción (no throttling por tiempo)
    if (Number(finalConfig.delay) === 0) {
      // Mantener el conteo de intentos y bloqueo si maxAttempts > 0
      if (!state) {
        const newState: ThrottleState = {
          lastSubmission: now,
          attemptCount: 1,
          windowStart: now,
          isBlocked: false,
        };
        this.throttleStates.set(actionId, newState);
        if (this.requiresPersistence(finalConfig)) {
          this.persistCriticalStates();
        }
        return true;
      }
      if (state.isBlocked) {
        return false;
      }
      // Reset the attempt count if the time window has passed
      if (now - state.windowStart > Number(finalConfig.timeWindow)) {
        state.attemptCount = 0;
        state.windowStart = now;
      }
      if (state.attemptCount >= Number(finalConfig.maxAttempts)) {
        state.isBlocked = true;
        if (this.requiresPersistence(finalConfig)) {
          this.persistCriticalStates();
        }
        setTimeout(() => {
          state.isBlocked = false;
          state.attemptCount = 0;
          state.windowStart = Date.now();
          if (this.requiresPersistence(finalConfig)) {
            this.persistCriticalStates();
          }
        }, Number(finalConfig.timeWindow));
        return false;
      }
      state.lastSubmission = now;
      state.attemptCount += 1;
      this.throttleStates.set(actionId, state);
      if (this.requiresPersistence(finalConfig)) {
        this.persistCriticalStates();
      }
      return true;
    }

    if (!state) {
      // First time this action is executed
      const newState: ThrottleState = {
        lastSubmission: now,
        attemptCount: 1,
        windowStart: now,
        isBlocked: false,
      };
      this.throttleStates.set(actionId, newState);
      // Persist if necessary
      if (this.requiresPersistence(finalConfig)) {
        this.persistCriticalStates();
      }
      return true;
    }

    // Check if the button is blocked
    if (state.isBlocked) {
      return false;
    }

    // Check the minimum delay between clicks
    if (now - state.lastSubmission < Number(finalConfig.delay)) {
      return false;
    }

    // Reset the attempt count if the time window has passed
    if (now - state.windowStart > Number(finalConfig.timeWindow)) {
      state.attemptCount = 0;
      state.windowStart = now;
    }

    // Check the attempt limit
    if (state.attemptCount >= Number(finalConfig.maxAttempts)) {
      state.isBlocked = true;
      // Persist the block immediately if necessary
      if (this.requiresPersistence(finalConfig)) {
        this.persistCriticalStates();
      }
      // Unblock after the time window
      setTimeout(() => {
        state.isBlocked = false;
        state.attemptCount = 0;
        state.windowStart = Date.now(); // Use current timestamp instead of stale 'now'
        // Persist the unblock if necessary
        if (this.requiresPersistence(finalConfig)) {
          this.persistCriticalStates();
        }
      }, Number(finalConfig.timeWindow));
      return false;
    }

    // Update state
    state.lastSubmission = now;
    state.attemptCount += 1;
    this.throttleStates.set(actionId, state);
    // Persist if necessary
    if (this.requiresPersistence(finalConfig)) {
      this.persistCriticalStates();
    }
    return true;
  }

  /**
   * Get the current throttling state for an action
   */
  getThrottleState(actionId: string): ThrottleState | null {
    return this.throttleStates.get(actionId) || null;
  }

  /**
   * Clear the throttling state for an action
   */
  clearThrottleState(actionId: string): void {
    const hadPersistentAction = this.shouldPersistAction(actionId);
    this.throttleStates.delete(actionId);

    // If it was a persistent action, update the storage
    if (hadPersistentAction) {
      this.persistCriticalStates();
    }
  }

  /**
   * Clear all throttling states
   */
  clearAllThrottleStates(): void {
    this.throttleStates.clear();
    // Limpiar también el storage
    localStorage.removeItem(this.storageKey);
  }

  /**
   * Clear the entire throttling storage (for debugging/admin)
   */
  clearPersistedStates(): void {
    localStorage.removeItem(this.storageKey);
    this.throttleStates.clear();
  }

  /**
   * Calculate the time remaining until the action is available again
   */
  getTimeUntilNextAction(actionId: string, config?: Partial<ThrottleConfig>): number {
    const finalConfig = { ...this.defaultConfig, ...config };
    const state = this.throttleStates.get(actionId);

    if (!state) return 0;

    const now = Date.now();
    const timeSinceLastSubmission = now - state.lastSubmission;
    const remainingDelay = Number(finalConfig.delay) - timeSinceLastSubmission;

    if (state.isBlocked) {
      const timeUntilUnblock = Number(finalConfig.timeWindow) - (now - state.windowStart);
      return Math.max(timeUntilUnblock, 0);
    }

    return Math.max(remainingDelay, 0);
  }

  /**
   * Get the number of attempts remaining before blocking
   */
  getRemainingAttempts(actionId: string, config?: Partial<ThrottleConfig>): number {
    const finalConfig = { ...this.defaultConfig, ...config };
    const state = this.throttleStates.get(actionId);

    // If the action is not in the state, return the max attempts
    if (!state) {
      return Number(finalConfig.maxAttempts);
    }

    // If the action is blocked, return 0
    if (state.isBlocked) {
      return 0;
    }

    // Reset the attempt count if the time window has passed
    const now = Date.now();
    if (now - state.windowStart > Number(finalConfig.timeWindow)) {
      return Number(finalConfig.maxAttempts);
    }

    // Return the remaining attempts
    return Math.max(Number(finalConfig.maxAttempts) - state.attemptCount, 0);
  }
}

// Export the class for testing
export { SecurityThrottleService };

// Singleton instance
export const securityThrottleService = new SecurityThrottleService();

/**
 * Genera un ID único para una acción basándose en el contexto
 */
export const generateActionId = (formName: string, actionType: string = 'submit'): string => {
  return `${formName}-${actionType}`;
};
