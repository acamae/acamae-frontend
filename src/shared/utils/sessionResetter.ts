let sessionTimeout: ReturnType<typeof setTimeout> | null = null;

/**
 * Reinicia el temporizador de sesión y ejecuta un callback tras el retraso especificado.
 * Útil para implementar cierre de sesión automático por inactividad.
 * @param callback Función a ejecutar cuando expira el temporizador
 * @param delay Tiempo en milisegundos antes de ejecutar el callback (por defecto 5000ms)
 */
export function resetSessionTimer(callback: () => void, delay: number = 5000) {
  if (sessionTimeout) {
    clearTimeout(sessionTimeout);
  }
  sessionTimeout = setTimeout(callback, delay);
}

/**
 * Limpia el temporizador de sesión actual, evitando la ejecución del callback.
 */
export function clearSessionTimer() {
  if (sessionTimeout) {
    clearTimeout(sessionTimeout);
    sessionTimeout = null;
  }
}
