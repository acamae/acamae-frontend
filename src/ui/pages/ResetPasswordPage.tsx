import React from 'react';
import { Card, Col, Row, Container } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';
import { useSearchParams } from 'react-router-dom';

import ResetPasswordForm from '@ui/components/Forms/ResetPasswordForm';

const ResetPasswordPage: React.FC = () => {
  const { t } = useTranslation();
  const [searchParams] = useSearchParams();
  const token = searchParams.get('token') || '';

  return (
    <Container className="py-5">
      <Row className="justify-content-center" data-testid="reset-password-page">
        <Col xs={12} md={8} lg={6} xl={5}>
          <Card>
            <Card.Header className="text-center">
              <h2 className="mb-4 text-center" data-testid="reset-password-title">
                {t('reset.title')}
              </h2>
            </Card.Header>
            <Card.Body>
              <ResetPasswordForm token={token} />
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default ResetPasswordPage;
