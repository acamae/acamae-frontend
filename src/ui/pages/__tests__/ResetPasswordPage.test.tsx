import { screen } from '@testing-library/react';
import { Routes, Route } from 'react-router-dom';

import { createTestProviderFactory } from '@shared/utils/renderProvider';
import ResetPasswordPage from '@ui/pages/ResetPasswordPage';

jest.mock('@ui/components/Forms/ResetPasswordForm', () => {
  const MockResetPasswordForm = () => (
    <div data-testid="mock-reset-password-form">Mocked Reset Password Form</div>
  );
  return MockResetPasswordForm;
});

function renderResetPasswordPage({ token = 'mock-token' }: { token?: string } = {}) {
  const renderWithProviders = createTestProviderFactory();
  return renderWithProviders(
    <Routes>
      <Route path="/restablecer-clave" element={<ResetPasswordPage />} />
    </Routes>,
    {
      route: `/restablecer-clave?${token ? `token=${token}` : ''}`,
    }
  );
}

describe('ResetPasswordPage', () => {
  it('should render the title', () => {
    renderResetPasswordPage();
    expect(screen.getByTestId('reset-password-title')).toBeInTheDocument();
  });

  it('snapshot render', () => {
    const { asFragment } = renderResetPasswordPage();
    expect(asFragment()).toMatchSnapshot();
  });
});
