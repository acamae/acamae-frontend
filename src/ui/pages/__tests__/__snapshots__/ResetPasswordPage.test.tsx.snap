// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordPage snapshot render 1`] = `
<DocumentFragment>
  <div
    class="py-5 container"
  >
    <div
      class="justify-content-center row"
      data-testid="reset-password-page"
    >
      <div
        class="col-xl-5 col-lg-6 col-md-8 col-12"
      >
        <div
          class="card"
        >
          <div
            class="text-center card-header"
          >
            <h2
              class="mb-4 text-center"
              data-testid="reset-password-title"
            >
              Restablecer contrase√±a
            </h2>
          </div>
          <div
            class="card-body"
          >
            <div
              data-testid="mock-reset-password-form"
            >
              Mocked Reset Password Form
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
